src/agent/custom-transform-agent.ts

```diff
@@ -0,0 +1,53 @@
+import {Expiry, HttpAgentRequest, HttpAgentRequestTransformFn} from '@dfinity/agent';
+import {isNullish, nowInBigIntNanoSeconds, uint8ArrayToBase64} from '@dfinity/utils';
+import {HexString} from '../types/hex-string';
+import {IcrcCallCanisterRequestParams} from '../types/icrc-requests';
+import {generateHash} from '../utils/crypto.utils';
+
+/**
+ * A custom transform function that processes the HTTP agent request.
+ *
+ * This transform function is intended to be used with the HttpAgent. It caches the expiry
+ * time of requests based on a hash of the request data and checks whether the cache is expired.
+ * If the cache is expired, it throws an error. If the cache is not expired, it reuses the cached
+ * expiry value.
+ *
+ * @returns {HttpAgentRequestTransformFn} The transform function that processes the request.
+ */
+export const customAddTransform = (): HttpAgentRequestTransformFn => {
+  const cache: Map<HexString, Expiry> = new Map();
+
+  return async (request: HttpAgentRequest) => {
+    const {canister_id, sender, method_name, arg, ingress_expiry, nonce} = request.body;
+
+    if (isNullish(nonce)) {
```

Comment:
Can you add a comment explaining why we don't do anything if no nonce is provided? We've iterated on this several times, so it might be good to document it for our future selves.